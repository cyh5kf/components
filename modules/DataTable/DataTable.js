import React,{Component} from "react"
import {Table,Icon,Checkbox,Button,Row,Col } from 'antd'
//import BaseForm,{FormItem} from 'components/BaseForm'


class DataTable extends Component{
  state={
    visible:false,
    newColumns:[],
    displayColumns:[]
  }
  static defaultProps={
    page:{},
    prefixCls: 'ant-table',
    showQuickJumper:true,
    pagination:{
       showTotal:total => `共 ${total} 条`,
       // showQuickJumper:true,
       showSizeChanger:true,
       pageSizeOptions:['10','20','50','100'],
    },
  //  scroll:{ y: 500 },
    style:{
      width:"100%"
    },
    showConfig:false,
    columns:[]
  }
  showPopover(){
    this.setState({
      visible:true
    })
  }
  constructor(props) {
    super(props)
    this.state.columns=props.columns
  }
  componentWillReceiveProps(nextProps){
    let {columns} = nextProps
      this.setState({
        columns:columns,
      });
  }

  onSelectChange(checkedValues){
    this.setState({
      columns:this.state.columns.map((col)=>{
        if(checkedValues.indexOf(col.key)>=0){
          col.visible=true
        }else{
          col.visible=false
        }
        return col
      })
    })
  }
  onClosePopup(){
    this.setState({
      visible:false
    })
  }
  onPopupVisibleChange(boolean){
     // console.log('show',arguments)
    this.setState({
      visible:boolean
    })
  }


  renderTableMenu(){
    // console.log("menu")
    let {columns}= this.state
    var defaultValue=columns.filter(col=>(col.type!='config' && col.visible==true)).map((col)=>col.key)
    // return (
    //     <div className="" style={{width:400,height:200,padding:'10px',border:'1px solid #cfdae5'}}>
    //       <TableMenu defaultValue={defaultValue} columns={columns} onSelectChange={this.onSelectChange.bind(this)} onClosePopup={this.onClosePopup.bind(this)} ></TableMenu>
    //     </div>
    //   )
    return (null)
  }
  render(){
    let {pagination,showConfig,page,...otherProps}= this.props
    let {visible,columns}=this.state
    let newColumns;
    if(showConfig){
      newColumns =columns.filter((col)=>{
        return col.visible==true
      }).concat([{
          title:" ",
          filterDropdown:(
            this.renderTableMenu()
          ),
          filterDropdownVisible:visible,
          onFilterDropdownVisibleChange:this.onPopupVisibleChange.bind(this),
          width:30,
          fixed:'right',
          type:'config'
      }])
    }else{
      newColumns=columns
    }
    return (<Table {...otherProps} columns={newColumns} pagination={!pagination?false:Object.assign({},pagination,page)}/>)
  }
}


export default DataTable
