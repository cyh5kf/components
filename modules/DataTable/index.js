import React,{Component} from 'react'

import Table from 'antd/lib/table'
import Icon from 'antd/lib/icon'
import Checkbox from 'antd/lib/checkbox'
import Button from 'antd/lib/button'
import Row from 'antd/lib/row'
import Col from 'antd/lib/col'
import Form from 'antd/lib/form'
//import BaseForm,{FormItem} from 'components/BaseForm'

class TableMenu extends Component{
  state = {
    visible: true,
    columns:[]
  }
  //请求远程数据接口
  componentWillMount() {
    let {actions} = this.props;

  }
  // //处理表格提交后动作
  handleOk(){
    const {columns} = this.state;
    const { onSelectChange,onClosePopup } =this.props
  //  console.log(columns)
    onSelectChange(columns)
  //  this.form.onSubmit()
    onClosePopup()
  }
  saveFormRef=(form)=>this.form=form
  handleSubmit(values){
    var {onSelectChange}=this.props
      this.setState({
        columns:values
      })
  //  console.log(values)
     // return new API().fetchTableColumns(values).then(json => {
     //   onSelectChange(values.isShowArr)
     //   // console.log(json,values)
     // }).catch(ex => {
     //   return "error"
     // })
  }
  handleChange(values){

    const {onSelectChange}=this.props
    this.setState({
      columns:values
    })
      // console.log(value)
    // const { onSelectChange } =this.props
    // onSelectChange(value)
  }
  render() {
    const {
      form,
      initialValues,
      handleSubmit,
      children,
      defaultValue,
      columns,
      onClosePopup
    } = this.props
    const saveFormRef=this.saveFormRef
    const formFullItemLayout = {
      labelCol: {
        span: 6
      },
      wrapperCol: {
        span: 18
      }
    }
    // onChange={this.onSelectChange.bind(this)}
  //  console.log(defaultValue,columns)
    return (
      <div className="" style={{width:400,height:200,padding:'10px',border:'1px solid #cfdae5',background:'#fff'}}>
        <Form onSubmit={handleSubmit} ref={saveFormRef} layout="inline">
            <Checkbox.Group name="isShowArr" style={{ width: '100%' }} defaultValue={defaultValue} onChange={this.handleChange.bind(this)}>
              <Row>
                {
                 columns.filter(it=>{
                   return it.title!='操作'
                 }).map((it,idx)=>{
                   return (<Col span={8} key={idx}><Checkbox value={it.key} disabled ={it.isRead==1?true:false} >{it.title}</Checkbox></Col>)
                 })
                }
              </Row>
            </Checkbox.Group>
            <div style={{textAlign:'right'}}>
             <Button size="small" onClick={onClosePopup}>取消</Button>
             <Button size="small" type="primary" onClick={this.handleOk.bind(this)} style={{marginLeft:'10px'}}>确定</Button>
            </div>
        </Form>
      </div>
    )
  }
}


class DataTable extends Component{
  state={
    visible:false,
    newColumns:[],
    displayColumns:[]
  }
  static defaultProps={
    page:{},
    prefixCls: 'ant-table',
    pagination:{
       showTotal:total => `共 ${total} 条`,
       // showQuickJumper:true,
       size:"middle",
       showSizeChanger:true,
       pageSizeOptions:['10','20','50','100'],
    },
  //  scroll:{ y: 500 },
    style:{
      width:"100%"
    },
    showConfig:false,
    columns:[]
  }
  showPopover(){
    this.setState({
      visible:true
    })
  }
  constructor(props) {
    super(props)
    this.state.columns=props.columns
  }
  componentWillReceiveProps(nextProps){
    let {columns} = nextProps
      this.setState({
        columns:columns,
      });
  }

  onSelectChange(checkedValues){
    //console.log(checkedValues)
    this.setState({
      columns:this.state.columns.map((col)=>{
        if(checkedValues.indexOf(col.key)>=0){
          col.visible=true
        }else{
          col.visible=false
        }
        return col
      })
    })
  }
  onClosePopup(){
    this.setState({
      visible:false
    })
  }
  onPopupVisibleChange(boolean){
     // console.log('show',arguments)
    this.setState({
      visible:boolean
    })
  }

  renderTableMenu(){
    let {columns}= this.state
    var defaultValue=columns.filter(col=>(col.type!='config' && (col.visible===true || col.visible===undefined))).map((col)=>col.key)
    return (
      <TableMenu defaultValue={defaultValue} columns={columns} onSelectChange={this.onSelectChange.bind(this)} onClosePopup={this.onClosePopup.bind(this)} ></TableMenu>
    )
  }
  render(){
    let {pagination,showConfig,page,...otherProps}= this.props
    let {visible,columns}=this.state
    let newColumns;
     if(showConfig){
    // if(true){
      newColumns =columns.filter((col)=>{
         return col.visible==true || col.visible==undefined
         // return true
      })
      // .concat([{
      //     title:" ",
      //     filterDropdown:(
      //       this.renderTableMenu()
      //     ),
      //     filterDropdownVisible:visible,
      //     onFilterDropdownVisibleChange:this.onPopupVisibleChange.bind(this),
      //     width:30,
      //     fixed:'right',
      //     type:'config'
      // }])
    }else{
      newColumns=columns
    }
    //console.log(newColumns,columns)
    return (<Table {...otherProps} columns={newColumns} pagination={!pagination?false:Object.assign({},pagination,page)}/>)
  }
}


export default DataTable
