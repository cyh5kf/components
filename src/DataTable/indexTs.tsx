import * as React from 'react';
import Table, {ColumnProps, TableProps} from 'antd/es/table';

interface ColumnsType {
  visible?: boolean;
  key: React.Key;
}

interface sortType {
  columnKey: string;
  order: string;
}

interface DataTableState<T> {
  visible: boolean;
  columns: ColumnProps<T>[];
  displayColumns: object[];
}

interface DataTableProps<T> extends TableProps<T> {
  page?: object;
  defaultSort?: sortType;
}

export default class DataTable<T> extends React.Component<DataTableProps<T>> {
  static defaultProps = {
    // page: {},
    prefixCls: 'ant-table',
    // pagination: {
    //   showTotal: (total: any) => `共 ${total} 条`,
    //   // showQuickJumper:true,
    //   size: 'middle',
    //   showSizeChanger: true,
    //   pageSizeOptions: ['10', '20', '50', '100']
    // },
    //  scroll:{ y: 500 },
    style: {
      width: '100%'
    },
    showConfig: false,
    columns: []
  };

  // constructor(props: CDataTableProps<T>) {
  //   super(props);
  //   this.state = {
  //     visible: false,
  //     displayColumns: [],
  //     columns: props.columns
  //   };
  // }
  // componentWillReceiveProps(nextProps: CDataTableProps<T>): void {
  //   const {columns} = nextProps;

  //   this.setState({
  //     columns: columns
  //   });
  // }
  // showPopover(): void {
  //   this.onPopupVisibleChange(true);
  // }
  // onClosePopup(): void {
  //   this.onPopupVisibleChange(false);
  // }
  // onPopupVisibleChange(bool: boolean): void {
  //   this.setState({
  //     visible: bool
  //   });
  // }

  // onSelectChange(checkedValues: React.Key[]): void {
  //   this.setState({
  //     columns: this.state.columns.map((col: any) => {
  //       if (checkedValues.indexOf(col.key) >= 0) {
  //         col.visible = true;
  //       } else {
  //         col.visible = false;
  //       }
  //       return col;
  //     })
  //   });
  // }
  render() {
    let {pagination, ...otherProps} = this.props;
    // let {columns} = this.state;
    // let newColumns = columns;
    // if (defaultSort !== undefined) {
    //   newColumns = columns.map(it => {
    //     defaultSort && defaultSort.columnKey == it.dataIndex
    //       ? (it = Object.assign(it, {defaultSortOrder: defaultSort.order}))
    //       : null;
    //     return it;
    //   });
    // }
    // let pagaSetting = pagination ? Object.assign({}, pagination, page) : false;
    return <Table {...this.props} />;
  }
}
