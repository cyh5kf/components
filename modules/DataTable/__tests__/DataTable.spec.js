import React from "react";
import { shallow } from "enzyme";
import toJson from "enzyme-to-json";
import { Table } from "antd";
import DataTable from "../index";

const setup = props => {
  // 模拟 props
  // const props = {
  //   // Jest 提供的mock 函数
  //   // onAddClick: jest.fn(),
  //   // actions:{
  //   //   listAction:jest.fn()
  //   // },
  //   reduce:{
  //     list:new Map(),
  //     spins:{
  //       tableSpin:false
  //     }
  //   }
  // }

  // 通过 enzyme 提供的 shallow(浅渲染) 创建组件
  const wrapper = shallow(<DataTable {...props} />);
  return {
    props,
    wrapper
  };
};

describe("DataTable 组件是否渲染", () => {
  const { wrapper, props } = setup();
  // case1
  // 通过查找存在 Input,测试组件正常渲染
  it("DataTable Component should be render", () => {
    //.find(selector) 是 Enzyme shallow Rendering 提供的语法, 用于查找节点
    // 详细用法见 Enzyme 文档 http://airbnb.io/enzyme/docs/api/shallow.html
    expect(wrapper.find("Table").exists()).toBe(true);
    // expect(toJson(wrapper)).toMatchSnapshot();
  });

  it.skip("DataTable Component should be pagination", () => {
    // expect(wrapper.find('Pagination').exists()).toBe(true);
  });
});

describe("state visible with some methods", () => {
  const { wrapper, props } = setup();

  it("default visible state is false", () => {
    expect(wrapper.state("visible")).toBe(false);
  });

  it("trggier showPopover visible to be true", () => {
    wrapper.instance().showPopover();
    expect(wrapper.state("visible")).toBe(true);
  });

  it("trggier onClosePopup visible to be false", () => {
    wrapper.instance().showPopover();
    wrapper.instance().onClosePopup();
    expect(wrapper.state("visible")).toBe(false);
  });

  it("trggier onPopupVisibleChange to any", () => {
    wrapper.instance().onPopupVisibleChange(true);
    expect(wrapper.state("visible")).toBe(true);
  });
});

describe("defaultSort ", () => {
  const columns = [
      {
        title: "first",
        key: 1,
        dataIndex: "first"
      },
      {
        title: "second",
        key: 2,
        dataIndex: "second"
      }
    ],
    defaultSort = {
      columnKey: "first",
      order: "descend"
    };
  const defaultProps = {
    rowKey: "id",
    defaultSort: defaultSort,
    dataSource: [
      {
        first: 111,
        second: 222,
        id: 1
      },
      {
        first: 333,
        second: 444,
        id: 2
      }
    ],
    columns: columns
  };

  it("while defaultSort is exist", () => {
    const { wrapper, props } = setup({
      ...defaultProps
    });
    expect(wrapper.state("columns")).toEqual(
      columns.map(it => {
        defaultSort.columnKey == it.dataIndex
          ? (it = Object.assign(it, { defaultSortOrder: defaultSort.order }))
          : null;
        return it;
      })
    );
  });

  it("while defaultSort null", () => {
    const { wrapper, props } = setup({
      ...Object.assign(defaultProps, { defaultSort: {} })
    });
    expect(wrapper.state("columns")).toEqual(columns);
  });
});
