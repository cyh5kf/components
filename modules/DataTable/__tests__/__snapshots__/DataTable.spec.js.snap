// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`快照测试 DataTable 快照测试 1`] = `
<Table
  bordered={false}
  className=""
  columns={Array []}
  dataSource={Array []}
  indentSize={20}
  loading={false}
  locale={Object {}}
  pagination={
    Object {
      "pageSizeOptions": Array [
        "10",
        "20",
        "50",
        "100",
      ],
      "showSizeChanger": true,
      "showTotal": [Function],
      "size": "middle",
    }
  }
  prefixCls="ant-table"
  rowKey="key"
  showHeader={true}
  size="default"
  style={
    Object {
      "width": "100%",
    }
  }
  useFixedHeader={false}
/>
`;
